src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -1004,12 +1043,21 @@ void FileSegmentsHolder::reset()
     ProfileEvents::increment(ProfileEvents::FilesystemCacheUnusedHoldFileSegments, file_segments.size());
     for (auto file_segment_it = file_segments.begin(); file_segment_it != file_segments.end();)
     {
-        /// One might think it would have been more correct to do `false` here,
-        /// not to allow background download for file segments that we actually did not start reading.
-        /// But actually we would only do that, if those file segments were already read partially by some other thread/query
-        /// but they were not put to the download queue, because current thread was holding them in Holder.
-        /// So as a culprit, we need to allow to happen what would have happened if we did not exist.
-        file_segment_it = completeAndPopFrontImpl(true);
+        try
+        {
+            /// One might think it would have been more correct to do `false` here,
+            /// not to allow background download for file segments that we actually did not start reading.
+            /// But actually we would only do that, if those file segments were already read partially by some other thread/query
+            /// but they were not put to the download queue, because current thread was holding them in Holder.
+            /// So as a culprit, we need to allow to happen what would have happened if we did not exist.
+            file_segment_it = completeAndPopFrontImpl(true);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+            chassert(false);
```

Comment:
Why do we need it?

Reply:
assert is to catch in CI, because there should be no errors.
try-catch is for safety just in case, because if there is an exception from one of the file segment's complete(), then we did not finish other file segments which could lead to let-over objects in cache memory.